### PR TITLE
[CSL-1901] Create tests for EpochPages, EpochSlots.

### PR DESCRIPTION
**Pull request by ksaric**
https://iohk.myjetbrains.com/youtrack/issue/CSL-1901

In the future we need to consider improving `generateValidExplorerMockableMode`.

### CODE REVIEW

File: explorer/src/Pos/Explorer/ExtraContext.hs [hunk 1]
```diff
@@ -172,3 +182,13 @@ instance (Monad m, MonadDBRead m, MonadSlotsData ctx m) =>
...
+
+    getEpochBlocksCSLI epoch page = do
+        extraCtx <- Ether.ask @ExtraContext
+        let explorerMockMode = ecExplorerMockableMode extraCtx
+        (emmGetEpochBlocks explorerMockMode) epoch page
+
+    getEpochPagesCSLI epoch = do
+        extraCtx <- Ether.ask @ExtraContext
+        let explorerMockMode = ecExplorerMockableMode extraCtx
+        emmGetEpochPages explorerMockMode $ epoch
```
Comment (sectore) [inline explorer/src/Pos/Explorer/ExtraContext.hs]: [Tiny] Missing newline

Comment (ksaric) [reply explorer/src/Pos/Explorer/ExtraContext.hs]: Fixed.

File: explorer/src/Pos/Explorer/ExtraContext.hs [hunk 2]
```diff
@@ -172,3 +182,14 @@ instance (Monad m, MonadDBRead m, MonadSlotsData ctx m) =>
...
```
Comment (akegalj) [inline explorer/src/Pos/Explorer/ExtraContext.hs]: > emmGetEpochPages explorerMockMode $ epoch

why `$` here ?

File: explorer/src/Pos/Explorer/TestUtil.hs [hunk 3]
```diff
@@ -101,6 +107,40 @@ generateValidExplorerMockableMode blocksNumber slotsPerEpoch = do
...
+              where
+                (as,bs) = splitAt n xs
+
+    createMapEpochMaxPages
+        :: (HasConfiguration)
+        => [EpochPagedBlocksKey]
+        -> Map Epoch Page
+    createMapEpochMaxPages epochPages = do
+        let groupedEpochPages :: [[(Epoch, Page)]]
+            groupedEpochPages = groupBy (\(a,_) (b,_) -> a == b) epochPages
```
Comment (akegalj) [inline explorer/src/Pos/Explorer/TestUtil.hs]: Use:
```
groupBy ((==) `on` fst))
```
`Data.Function.on`

File: explorer/src/Pos/Explorer/TestUtil.hs [hunk 4]
```diff
@@ -101,6 +107,40 @@ generateValidExplorerMockableMode blocksNumber slotsPerEpoch = do
...
+
+        fromList $ maximumBy (\(a,_) (b,_) -> compare a b) <$> groupedEpochPages
```
Comment (akegalj) [inline explorer/src/Pos/Explorer/TestUtil.hs]: Use:
```
maximumBy (compare `on` fst)
```
`Data.Function.on`

File: explorer/src/Pos/Explorer/ExtraContext.hs (hunk 2 above)
Comment (akegalj) [reply explorer/src/Pos/Explorer/ExtraContext.hs]: The same applies for others in the same module

Comment (akegalj) [reply explorer/src/Pos/Explorer/ExtraContext.hs]: Also, there is a lot of copy paste in this module. I see 
```
+        extraCtx <- Ether.ask @ExtraContext
+        let explorerMockMode = ecExplorerMockableMode extraCtx
```
being used at 7 places

Comment (ksaric) [reply explorer/src/Pos/Explorer/ExtraContext.hs]: That's two lines on 7 places - hardly a lot. 
I saw that, and understand why you would comment that, but I decided that it's more understandable this way, I really don't think we should replace 7-14 lines of code with something that's less understandable.
I really think there is a fine line between removing lines of code and making something easier to read.

File: explorer/src/Pos/Explorer/TestUtil.hs (hunk 3 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/TestUtil.hs]: Useful, will fix.

File: explorer/src/Pos/Explorer/TestUtil.hs (hunk 4 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/TestUtil.hs]: Useful, will fix.

File: explorer/src/Pos/Explorer/ExtraContext.hs (hunk 2 above)
Comment (akegalj) [reply explorer/src/Pos/Explorer/ExtraContext.hs]:
```haskell
someFunctionName someVariable = ecExplorerMockableMode <$> Ether.ask @ExtraContext
```
how is this confusing?

>  it's more understandable this way, I really don't think we should replace 7-14 lines of code with something that's less understandable. I really think there is a fine line between removing lines of code and making something easier to read.

and more error prone and harder to maintain

Comment (ksaric) [reply explorer/src/Pos/Explorer/ExtraContext.hs]: @akegalj Why don't you try writing that function? And compiling? If you succeed, you can push the changes, but it's hard for me to explain something if you didn't try it yourself.

File: explorer/src/Pos/Explorer/TestUtil.hs (hunk 4 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/TestUtil.hs]: Fixed.

File: explorer/src/Pos/Explorer/TestUtil.hs (hunk 3 above)
Comment (ksaric) [reply explorer/src/Pos/Explorer/TestUtil.hs]: Fixed.

**Comment by ksaric** [comment]
@akegalj Fixed the review changes you requested except for one. Left a comment there.

**Pull request merged**